scipy/interpolate/tests/test_interpolate.py

```diff
@@ -700,6 +700,20 @@ def test_spline_nans(self):
                 assert_(np.isnan(outn).all())
                 assert_equal(out.shape, outn.shape)
 
+    def test_all_nans(self):
+        # regression test for gh-11637: interp1d core dumps with all-nan `x`
+        x = np.ones(10) * np.nan
+        y = np.arange(10)
+        with assert_raises(ValueError):
+            interp1d(x, y, kind='cubic')
+
+        # leading nans are OK (implementation detail: python min() mishandles)
+        x = np.arange(10)
+        y = np.arange(10)
+        x[:2] = np.nan
+        with assert_raises(ValueError):
+            interp1d(x, y, kind='cubic')
```

Comment:
I'm a little confused here--the comment above says "leading nans are OK," but you are asserting for a `ValueError` in this case where not all values are `nan`?

The comment above the source code you changed also indicates:
```python
# Quadratic or cubic spline. If input contains even a single
# nan, then the output is all nans.
```

That seems to suggest returning `nan`-filled array instead of raising?

Reply:
> That seems to suggest returning nan-filled array instead of raising?

Exactly. 

With this PR:


```
In [1]: import numpy as np                                                      

In [2]: from scipy.interpolate import interp1d                                  

In [3]: interp1d([np.nan, 1, 2, 3, 4], np.arange(5), kind='cubic')([1, 2, 3])   
/home/br/repos/scipy/build/testenv/lib/python3.5/site-packages/scipy/interpolate/interpolate.py:694: RuntimeWarning: invalid value encountered in greater
  above_bounds = x_new > self.x[-1]
Out[3]: array([nan, nan, nan])

In [4]: interp1d([np.nan]*5, np.arange(5), kind='cubic')([1, 2, 3])             
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-db3be8389b71> in <module>
----> 1 interp1d([np.nan]*5, np.arange(5), kind='cubic')([1, 2, 3])

~/repos/scipy/build/testenv/lib/python3.5/site-packages/scipy/interpolate/interpolate.py in __init__(***failed resolving arguments***)
    527                     sx = self.x[~mask]
    528                     if sx.size == 0:
--> 529                         raise ValueError("`x` array is all-nan")                        
    530                     xx = np.linspace(np.nanmin(self.x),
    531                                      np.nanmax(self.x),

ValueError: `x` array is all-nan
```

On master cell [4] would segfault, and that's what this PR fixes. Cell [3] is identical on master and this branch.

Reply:
In fact, that part of the test was wrong and is gone in the force-push.

(The intent of the test was already covered)

Reply:
Why is erroring out the desired behavior here? Wouldn't it be better to fill the array with zeros, then get the right shape, and return all-nans? That way `[np.nan, np.nan, 1.61]` and `[np.nan, np.nan, np.nan]` are consistent.

Reply:
Well, it's a corner case. Interpolation is within a closed interval, and what is the interpolation range with less then two `x` non-nan values? So there's an interaction with `bounds_error`.
The current behavior with a single non-nan value is not very sane either:

```
In [1]: import numpy as np                                                      

In [2]: from scipy.interpolate import interp1d                                  

In [3]: f = interp1d([np.nan, np.nan, np.nan, np.nan, 1.6], [1, 2, 3, 4, 5])    

In [6]: f(-1)                                                                   
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-8e6746a6b70d> in <module>
----> 1 f(-1)

~/repos/scipy/build/testenv/lib/python3.5/site-packages/scipy/interpolate/polyint.py in __call__(self, x)
     77         """
     78         x, x_shape = self._prepare_x(x)
---> 79         y = self._evaluate(x)
     80         return self._finish_y(y, x_shape)
     81 

~/repos/scipy/build/testenv/lib/python3.5/site-packages/scipy/interpolate/interpolate.py in _evaluate(self, x_new)
    661         y_new = self._call(self, x_new)
    662         if not self._extrapolate:
--> 663             below_bounds, above_bounds = self._check_bounds(x_new)
    664             if len(y_new) > 0:
    665                 # Note fill_value must be broadcast up to the proper size

~/repos/scipy/build/testenv/lib/python3.5/site-packages/scipy/interpolate/interpolate.py in _check_bounds(self, x_new)
    690         # !! Could provide more information about which values are out of bounds
    691         if self.bounds_error and below_bounds.any():
--> 692             raise ValueError("A value in x_new is below the interpolation "
    693                              "range.")
    694         if self.bounds_error and above_bounds.any():

ValueError: A value in x_new is below the interpolation range.

In [7]: f(10)                                                                   
Out[7]: array(nan)
```

Segfaults should not happen, even in this weird corner cases, but other than that I'm not sure it's worth spending effort on.

Reply:
OK, I did step it in pdb. With `x=[nan, nan, nan, 1.6]` and `bounds_error=True` (default) and `assume_sorted=False` (default), numpy sorts nans to the end of the array, so that `self.x` becomes `array([1.6, nan, nan, nan, nan])`. And then in this check

```
686  	        # which values are outside the boundary region.
687  ->	        below_bounds = x_new < self.x[0]
688  	        above_bounds = x_new > self.x[-1]
```

for `f(-1)` it correctly infers that -1 is below bounds, but for `f(2)` it does not, because it compares `2 > nan` and that one is False.

Conclusion: it's a mess. Cleaning it up is a bit of a hassle and is fraught with back compat issues. If it errors out for now, we'll be able to fix it, if/when a use case materializes.

Reply:
Okay, thanks for looking into it. Agreed, if it's complex to change that not worth spending time on.